commit e556ce2c9d2163aea122e91a1512c9a110fece5d
Author: Yan Zheng <zheng.yan@oracle.com>
Date:   Thu Nov 20 10:25:19 2008 -0500

    Btrfs: Drop dirty roots created by log replay immediately when
    
    The log replay produces dirty roots. These dirty roots
    should be dropped immediately if the fs is mounted as
    ro. Otherwise they can be added to the dirty root list
    again when remounting the fs as rw. Thank you,
    
    Signed-off-by: Yan Zheng <zheng.yan@oracle.com>

diff --git a/fs/btrfs/disk-io.c b/fs/btrfs/disk-io.c
index 9816522..fb04665 100644
--- a/fs/btrfs/disk-io.c
+++ b/fs/btrfs/disk-io.c
@@ -1757,6 +1757,11 @@ struct btrfs_root *open_ctree(struct super_block *sb,
 						      generation + 1);
 		ret = btrfs_recover_log_trees(log_tree_root);
 		BUG_ON(ret);
+
+		if (sb->s_flags & MS_RDONLY) {
+			ret =  btrfs_commit_super(tree_root);
+			BUG_ON(ret);
+		}
 	}
 
 	if (!(sb->s_flags & MS_RDONLY)) {

